### ISSUE TITLE
Allow for caching of Username and PW with GitHub Package

### ISSUE THREAD

**Issue by jcastle** [atom/github#861]
Each time you go to interact with a GitHub repository through Atom, you are required to enter Username and PW.  For those using a token due to 2FA, this is painful.  Please allow for caching/saving user login information.

![screen shot 2017-05-23 at 5 47 52 pm](https://cloud.githubusercontent.com/assets/5223678/26377829/f7455ed4-3fdf-11e7-8651-935d7296333a.png)

**Comment by smashwilson** [comment]
Do you have a [credential helper](https://help.github.com/articles/caching-your-github-password-in-git/) installed for git?

If not, for the moment that's what we're recommending for credential caching, because Atom will be able to use cached credentials entered at the command line and vice versa. In this case we definitely have a usability problem, though. I've contemplated either implementing a "remember me" checkbox or linking out to that help article from the credential dialog itself as a kind of hint.

If you do, then there's something wrong with the way we interact with the credential helpers! To help us diagnose, can you enable "git diagnostics" in the package settings (via the settings view), trigger a fetch, push, or pull, then expand the appropriate git command in your developer console? That will run git commands with `GIT_TRACE=1` and enable a lot of logging that'll be helpful for seeing why it isn't working for you.

/cc @mauriciord who's reported encountering a similar issue in #436.

**Comment by jcastle** [comment]
@smashwilson, took me a while to check this out.  Been using terminal in the meantime.  Have credential helper installed and ran some tests on Atom 1.18 with it.  Still didn't change the login process to push files to GitHub.  Not a big deal to regenerate the token as PW but still easier to use terminal than to generate on a fresh start.  I use terminal for file manipulation so usually have open.  Ultimately need a way to store credentials in Atom/Git.  

Thanks for you feedback, def a great product, keep it up!

**Comment by smashwilson** [comment]
@jcastle Hmm. Can you try the `PATH` workaround in #436?

**Comment by Bohdan-Khomtchouk** [comment]
@smashwilson Hi!  Same problem here as @jcastle described above.  I tried the `PATH` workaround in #436 but no luck.  Really looking forward to a `Remember Github credentials` button :)  Meanwhile, will have to suffer with the username/password copy-pasta at each push/pull.

**Comment by rcrooks** [comment]
@Bohdan-Khomtchouk did you also add the command `export PATH=${PATH}:$(git --exec-path)` to your ~/.bash_profile file (and reboot)? That finally did it for me, after I'd tried all the other suggestions/steps.

**Comment by jonnylink** [comment]
Just to review the ticket history here:
- Ticket 861 closed as a duplicate of ticket 1033
- Ticket 1033 closed as a duplicate of ticket 861

Maybe I'm missing something. How is this helpful @rsese?

**Comment by Ben3eeE** [comment]
@jonnylink #861 (this issue) is not closed.

**Comment by 50Wliu** [comment]
@jonnylink Apologies for the confusion.  Maybe it would be helpful if you could point out when this issue was previously closed, as it appears to me that this issue hasn't been closed since it was opened.

**Comment by AGraber** [comment]
I join this issue as it's a pain to rewrite credentials over and over.
I hope we can get something here

**Comment by jonnylink** [comment]
@50Wliu ....it wasn't. I'm a dope. Sorry.

**Comment by 50Wliu** [comment]
No problem - happens sometimes, given the amount of issues :).

Have you tried @smashwilson's suggestions in https://github.com/atom/github/issues/861#issuecomment-303553662?

**Comment by OctoPenguin** [comment]
Why not do a GitHub OAuth process so you wouldn't have to do this stupid authenticating thing over and over and over?

**Comment by raphalvessa** [comment]
@OctoPenguin That's fine, and I support your idea, but it wouldn't work with Bitbucket or other services. You need to think of a universal Git solution.

**Comment by OctoPenguin** [comment]
Well, since GitHub boasts it's integration with Atom, just continue boosting. For other Git services, maybe you could have a built-in credential caching function.

**Comment by rgawenda** [comment]
What I don't get is:
1. Github will happily accept my ssh key from a git CLI commit
2. Atom can use and uses that key from some packages
3. 1+2=

**Comment by Ruben9922** [comment]
I tried running `git config --global credential.helper wincred` as per [this help page](https://help.github.com/articles/caching-your-github-password-in-git/) someone linked near the top of this thread. I also tried `git config --global credential.helper cache` as mentioned by someone in [this duplicate issue](https://github.com/atom/github/issues/1107). Neither seemed to work (the first command actually did seem to work but only until I restarted Atom for some reason).

The following seemed to fix the problem for me on Windows:
* Go to *Control Panel* > *User Accounts* > *Credential Manager*
* Click the *Windows Credentials* button
* Under *Generic Credentials* there was a credential called `atom-github/https://api.github.com`
* Edit this credential to contain your GitHub username and password.
  (For me previously it only contained a password.)

![image](https://user-images.githubusercontent.com/14931620/30112706-27f372aa-930a-11e7-97dd-99816005ab4c.png)

Maybe the fact the credential only initially contained a password indicates that the GitHub plugin in Atom isn't storing the credentials correctly.

**Comment by ItsPugle** [comment]
I don't know the conversation so far, but for me (1.20.0 on macOS), Atom doesn't remember login credentials regardless of any previous pushes. Not only that but even with the credential helper installed, I'm still prompted everytime. 

I don't get prompted when I use the Terminal though.

**Comment by zchsh** [comment]
Was also getting prompted for credentials every time I hit push, and usual username and password were being rejected, which I guess is expected with 2FA ... I've generated a personal access token, entered that as my 'password', and followed instructions on https://help.github.com/articles/caching-your-github-password-in-git/ and things seem to be working now. But would love if that process was a bit more intuitive!

**Comment by Selindrile** [comment]
Is there any movement on this? Specifically the remember me button >.>. A simple solution would be very nice for those of us not terribly savvy at such things :/

**Comment by sivaneshl** [comment]
For me, atom doesn't ask for credentials in my windows10 machine, but in my ubuntu it does

**Comment by Hamberfim** [comment]
Storing credentials in Atom through a check box would be great.  Switching back and forth between Atom and the terminal or re-entering credentials in Atom on every push is getting to be a bit of an annoyance.

**Comment by Mei152** [comment]
I've been switching to terminal to do pushes because of this issue too.
It would be great to have Atom remember the credentials.

**Comment by ChrisAtBrookes** [comment]
This is the one issue that is preventing me using Github with my students in a lab of MACs. Although I can use the credential helper as admin an a MAC, the students don't have that privilege.

I appreciate that they should be using command line via the terminal, but this is for creating a web site and delivering in GitHub pages all via github classroom. Strange that *Dreamweaver* now uses GitHub **and** saves the username / password. Also *strange* that *GitHub Desktop* saves the credentials as well.

**Comment by evansharp** [comment]
I'll echo @ChrisAtBrookes about the strangeness of GH Desktop being able to do this but no Atom package for it. 
I thought I was losing my mind today setting up a new dev laptop: trying to figure out where I had entered a personal token on my iMac to make it not prompt me for creds, then recalling that I have GH Desktop installed there. Since there is no Desktop client for Linux, I'm stuck with this issue on the laptop.

I'm also seeking a solution to streamline this for CS students. Having them enter creds every push just doesn't fly with teenagers; not that it really flys with me either.

Does anyone know why the Github integration panel accepts a token but does not apply it within the Git panel?

**Comment by timfallmk** [comment]
Hi all. Since I still see activity on this thread, and for anyone coming here from Google, this is due to a misconfiguration in Atom over the path to the installed version of `git`.

**tl;dr** [Follow the process here](https://github.com/atom/github/issues/861#issuecomment-314258189)

**Longer**: The issue stems from Atom not using the built-in `git` executable, and it's associated configuration. Since a system `git` is typically configured to use one or more credential helpers, it must know to delegate to a credential helper when authentication is requested. A slight change in the way Atom detects the paths for `git` (and it's configuration) in the system caused it to fall back to the default behavior, which is to prompt for credentials on every request. With a properly configured path to the correct `git`, this doesn't happen.

For me, an update to the Canary branch of Atom around August of 2017 corrected the original issue. For manual fixes see [this part of the thread](https://github.com/atom/github/issues/861#issuecomment-314258189).

**Hopefully this closes this issue for everyone having this problem and ending up here**

**Comment by andrewhharmon** [comment]
I'm having this same problem on Windows. Tried to follow the process linked to above but could not get it to work. git bash for windows is working fine as well as other 3rd party tools like Eclispe. But Atom requests username and password every time. Any fixes for windows users?

**Comment by BoBBer446** [comment]
Why is it so hard to intigate? It's terrible that you have to re-enter the credentials each time. I doubt that the developers themselves work with ATOM or use this function. Otherwise, it'd probably be intigated by now. That's too bad.

**Comment by andrewhharmon** [comment]
I’ve been using Atom on windows for about a year with flawless github integration. This issue only recently popped up I think after an atom update. Hoping it’s a quick bug fix or some settings need to be tweaked

**Comment by rcrooks** [comment]
since following the instructions higher in this thread, I've had no issues (I'm on a Mac though, so don't know how well it works on Windows)

**Comment by BoBBer446** [comment]
It's strange that it takes such a detour to do such a simple thing.

**Comment by andrewhharmon** [comment]
and to clarify, I'm not using two factor authentication. Just regular old un/pw and it's still an issue. I see above the short term road map mentions 2FA so not sure if the issue is specifically related to that.

**Comment by andrewhharmon** [comment]
well i have this working now. For what it's worth, I went into Control Panel\User Accounts\Credential Manager and in Windows Credentials i delete any credentials relating to github.  I issued this command 
`git config --global credential.helper wincred` Then in git bash i did a pull and it prompted me for un/pw. I supplied it and it created new entries in Windows Credentials. I ran another pull in git bash and did not get the un/pw prompt so i knew my creds were stored.  I then opened Atom and did a pull and it worked fine with no prompt for un/pw.  Hope that helps anyone else out there

**Comment by rsese** [comment]
@BoBBer446 your comment was deleted as a violation of the [Atom Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md) as it was insulting or derogatory. You may consider this an official warning.

**Comment by BoBBer446** [comment]
It is of course easy to decide for yourself what should be censored / deleted instead of leaving it to the public. I didn't use swearwords or insult anyone. It's a real pity such a childish attitude here.

**Comment by ShtHappens796** [comment]
This issue is here in order to keep track of a problem and help resolving it. Spamming our mailboxes with what seems to be a meaningless fight does not help the cause at all. 

The developers of Atom are coding for the community and they don't get paid for their efforts, so let them do what they do best, contributing to an open source project. Being impatient does not help you or the community.

Thanks for understanding.

**Comment by brunotdantas** [comment]
Hello Guys the solution i did was preety much the same as @andrewhharmon.
1. Go to Windows credentials
2. Check all the credentials that are stored for ATOM/GITHUB
3. Replace them if needed or complete if any of them are blank
4. open CMD
5. type :git config --global credential.helper wincred
6. close and open atom again and push something to GITHUB

This worked just fine to me :+1:

**Comment by Dan9er** [comment]
@andrewhharmon @brunotdantas What about Ubuntu?

**Comment by andrewhharmon** [comment]
ooh, you got me there. Not familiar with how Ubuntu stores these creds

**Comment by sguergachi** [comment]
I'm having this issue on Windows and it's very frustrating, since it worked perfectly not long ago.
Every time I try fetching, pulling, or pushing using the atom github package I have to put in my credentials.

I don't have 2FA, I followed @brunotdantas instructions (all github/atom related credentials filled + CMD command) and I still have the same issue.

**Comment by andrewhharmon** [comment]
did you see my post above? I had to delete (not edit) the credentials for it to work.

**Comment by sguergachi** [comment]
@andrewhharmon That worked! Thanks!

**Comment by smashwilson** [comment]
#1327 adds a "remember" checkbox to the credential handler dialog. Check it and your credentials will be stored and retrieved from your OS keychain, which should alleviate this pain considerably :old_key: 

The credential helper still takes pains to use git's configured credential helpers if you have any installed, and you may wish to use one regardless. Credentials you store there will be used from both Atom and the command line (or other git tooling!), but credentials you save in Atom will only be used in Atom.

As an added enhancement, if you've logged in to the GitHub panel, that token will be used to auth. I wasn't sure if it had enough permissions but it seems so :tada:

**Comment by andrewhharmon** [comment]
Cool, Thanks

**Comment by schollii** [comment]
For me what worked is turning on credential caching in git, and setting the timeout (which is in seconds) to a high value: 
```ini
[credential]
	helper = cache --timeout 1000000
```
The above that if I don't commit for more than 11.5 days I will have to re-enter the password; I can live with that :)

Caching has the advantage that password is not stored in a file, only in memory.
